### PR TITLE
Set train and eval for criterion and metric if appropriate

### PR DESCRIPTION
**Pull request by bstriner**
Set train and eval for model, criterion, and metric (if the latter 2 exist and are Modules).

Example use case is I've got cross entropy with smoothing. I want to smooth the loss during training, but I don't want to apply smoothing during testing.

Called the functions `eval_mode` and `train_mode` so it is obvious that the function doesn't actually do any training or evaluating.

### CODE REVIEW

File: inferno/trainers/basic.py [hunk 1]
```diff
@@ -422,6 +422,22 @@ def metric_is_defined(self):
         """Checks if the metric is defined."""
         return self._metric is not None
 
+    def eval_mode(self):
+        """Set model, criterion and metric to eval mode"""
+        self.model.eval()
+        if self.criterion_is_defined and isinstance(self.criterion, torch.nn.Module):
+            self.criterion.eval()
+        if self.metric_is_defined and isinstance(self.metric, torch.nn.Module):
+            self.metric.eval()
```
Comment (nasimrahaman) [inline inferno/trainers/basic.py]: A `return self` here and after line 440, and we're good to go.

**Comment by nasimrahaman** [comment]
Thanks!
